File: bowling/Bowling08-SpecFlow/Bowling08-SpecFlow/Game.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Bowling08_SpecFlow
{
    class Game
    {
        private readonly IList<int> _rolls = new List<int>();
        
        public void Roll(int pinsDown)
        {
            _rolls.Add(pinsDown);
        }

        public int GetScore()
        {
            int score = 0;
            int i = 0;
            
            for (int frameIndex = 0; frameIndex < 10; frameIndex++)
            {
                if (FrameIsStrike(i))
                {
                    score += AddStrikeScore(i);
                    i++;
                }
                else if (FrameIsSpare(i))
                {
                    score += AddSpareScore(i);
                    i += 2;
                }
                else
                {
                    score += AddTypicalFrameScore(i);
                    i += 2;
                }
            }
            
            return score;
        }

        private int AddTypicalFrameScore(int i)
        {
            return _rolls[i] + _rolls[i + 1];
        }

        private int AddSpareScore(int i)
        {
            return 10 + _rolls[i+2];
        }

        private int AddStrikeScore(int i)
        {
            return 10 + _rolls[i + 1] + _rolls[i + 2];
        }

        private bool FrameIsSpare(int i)
        {
            return _rolls[i] + _rolls[i + 1] == 10;
        }

        private bool FrameIsStrike(int i)
        {
            return _rolls[i] == 10;
        }
    }
}

File: bowling/Bowling10/src/Game.java
public class Game {

    int[] rolls = new int[21];
    int rollIndex = 0;

    public void roll(int pinsDown) {
        rolls[rollIndex++] = pinsDown;
    }

    public int getScore() {

        int score = 0;
        int rollCounter= 0;

        for (int frameIndex = 0; frameIndex < 10; frameIndex++) {

            if(currentFrameIsStrike(rolls[rollCounter])) {
                score += getStrikeFrameScore(rollCounter);
                rollCounter++;
            }
            else if(currentFrameIsSpare(rollCounter)) {
                score += getSpareFrameScore(rollCounter);
                rollCounter += 2;
            }
            else {
                score += getStandardFrameScore(rollCounter);
                rollCounter += 2;
            }
        }

        return score;
    }

    private boolean currentFrameIsSpare(int rollCounter) {
        return rolls[rollCounter] + rolls[rollCounter+1] == 10;
    }

    private boolean currentFrameIsStrike(int roll) {
        return roll == 10;
    }

    private int getStrikeFrameScore(int rollCounter) {
        return 10 + rolls[rollCounter +1] + rolls[rollCounter +2];
    }

    private int getStandardFrameScore(int rollCounter) {
        return rolls[rollCounter] + rolls[rollCounter + 1];
    }

    private int getSpareFrameScore(int rollCounter) {
        return 10 + rolls[rollCounter+2];
    }
}
